### PR TITLE
Human prefs base

### PR DESCRIPTION
**Pull request by mschweizer**
## Description

Refactored gatherer / querent class structure

### CODE REVIEW

File: src/imitation/algorithms/preference_comparisons.py [hunk 1]
```diff
@@ -820,30 +820,27 @@ def __call__(
...
-        custom_logger: Optional[imit_logger.HierarchicalLogger] = None,
+            self,
+            video_output_dir: str,
+            video_type="webm",
+            video_fps: int = 20,
+            rng: Optional[np.random.Generator] = None,
+            custom_logger: Optional[imit_logger.HierarchicalLogger] = None,
     ):
-        """Initializes the PrefCollect querent.
+        """Initializes the querent.
```
Comment (rk1a) [inline src/imitation/algorithms/preference_comparisons.py:836]: Add some class details.

File: src/imitation/algorithms/preference_comparisons.py [hunk 2]
```diff
@@ -820,30 +820,27 @@ def __call__(
...
 
         Args:
-            pref_collect_address: end point of the PrefCollect web service.
             video_output_dir: path to the video clip directory.
```
Comment (rk1a) [inline src/imitation/algorithms/preference_comparisons.py:839]: `video_type` missing

File: src/imitation/algorithms/preference_comparisons.py [hunk 3]
```diff
@@ -854,87 +851,92 @@ def __call__(
...
             self._query(query_id)
-
         return identified_queries
 
-    def _query(self, query_id):
-        requests.put(
-            self.query_endpoint + query_id,
-            json={"uuid": "{}".format(query_id)},
+    def _write_query_videos(self, query_id, query):
+        output_file_name = os.path.join(
```
Comment (rk1a) [inline src/imitation/algorithms/preference_comparisons.py]: use `pathlib` for definition of path.
```
output_file_name = Path(self.video_output_dir) /  f"{query_id}" + "-{}" + f".{self.video_type}"
```

File: src/imitation/algorithms/preference_comparisons.py [hunk 4]
```diff
@@ -854,87 +851,92 @@ def __call__(
...
+
+        Args:
+            collection_service_address: Network address of the collection service's REST interface.
+            video_output_dir: path to the video clip directory.
+            video_fps: frames per second of the generated videos.
+            rng: random number generator, if applicable.
+            custom_logger: Where to log to; if None (default), creates a new logger.
+        """
+        super().__init__(video_output_dir, video_fps=video_fps, rng=rng, custom_logger=custom_logger)
+        self.query_endpoint = collection_service_address + "/preferences/query/"
```
Comment (rk1a) [inline src/imitation/algorithms/preference_comparisons.py]: Make route configurable. Either include in endpoint or define new attribute.

File: src/imitation/scripts/config/train_preference_comparisons.py [hunk 5]
```diff
@@ -92,7 +92,7 @@ def synch_human_preferences():
 
 @train_preference_comparisons_ex.named_config
 def human_preferences():
-    gatherer_cls = preference_comparisons.PrefCollectGatherer
+    gatherer_cls = preference_comparisons.RESTGatherer
     gatherer_kwargs = dict(
         pref_collect_address="http://127.0.0.1:8000",
```
Comment (rk1a) [inline src/imitation/scripts/config/train_preference_comparisons.py]: Change to correct attribute name.

File: src/imitation/algorithms/preference_comparisons.py [hunk 6]
```diff
@@ -854,87 +851,92 @@ def __call__(
...
+        """Override this method in subclasses to specify query behavior."""
+        pass
+
+
+class RESTQuerent(VideoBasedQuerent):
+    """Sends queries to a REST web service."""
+
+    def __init__(
+        self,
+        collection_service_address: str,
```
Comment (rk1a) [inline src/imitation/algorithms/preference_comparisons.py:949]: Is there a special type for rest endpoints?

File: src/imitation/algorithms/preference_comparisons.py [hunk 7]
```diff
@@ -854,87 +851,92 @@ def __call__(
...
+        video_output_dir: str,
```
Comment (rk1a) [inline src/imitation/algorithms/preference_comparisons.py:950]: standard type is `str | os.PathLike`

File: src/imitation/algorithms/preference_comparisons.py [hunk 8]
```diff
@@ -854,87 +851,92 @@ def __call__(
...
+            collection_service_address: Network address of the collection service's REST interface.
+            video_output_dir: path to the video clip directory.
+            video_fps: frames per second of the generated videos.
+            rng: random number generator, if applicable.
+            custom_logger: Where to log to; if None (default), creates a new logger.
+        """
+        super().__init__(video_output_dir, video_fps=video_fps, rng=rng, custom_logger=custom_logger)
+        self.query_endpoint = collection_service_address + "/preferences/query/"
+
+    def _query(self, query_id):
```
Comment (rk1a) [inline src/imitation/algorithms/preference_comparisons.py]: Add docstring describing the REST interface. The expected query interface.

File: src/imitation/algorithms/preference_comparisons.py [hunk 9]
```diff
@@ -854,87 +851,106 @@ def __call__(
...
-                "Unsupported observation type "
-                f"for writing fragment video: {type(fragment.obs)}",
-            )
-    # Note: `ImageSeqeuenceClip` handily accepts both
-    #       lists of image paths or numpy arrays
-    clip = ImageSequenceClip(frames_list, fps=frames_per_second)
-    moviepy_logger = None if not progress_logger else "bar"
-    clip.write_videofile(output_path, logger=moviepy_logger)
+    def _load_video_data(self, query_id):
+        video_data = []
```
Comment (rk1a) [inline src/imitation/algorithms/preference_comparisons.py]: To be tested

File: src/imitation/algorithms/preference_comparisons.py [hunk 10]
```diff
@@ -1298,33 +1303,15 @@ def is_running_pytest_test(self) -> bool:
         return "PYTEST_CURRENT_TEST" in os.environ
 
 
-class PrefCollectGatherer(PreferenceGatherer):
-    """Gathers preferences from PrefCollect interface."""
-
+class AsynchronousHumanGatherer(PreferenceGatherer, abc.ABC):
```
Comment (rk1a) [inline src/imitation/algorithms/preference_comparisons.py]: Maybe combine with `SynchronousHumanGatherer`. e.g. by checking whether all queries have been answered before proceding in the `gather()` call.

File: src/imitation/algorithms/preference_comparisons.py
Comment (rk1a) [inline src/imitation/algorithms/preference_comparisons.py:1]: Rename `AsynchronousHumanGatherer` to `HumanGatherer`; motivation synchronous and asynchronous gathering can be combined.
Rename `SynchronousHumanGatherer` to `NotebookGatherer` (or something like that).

File: src/imitation/algorithms/preference_comparisons.py [hunk 11]
```diff
@@ -1350,6 +1337,39 @@ def gather(self) -> Tuple[Sequence[TrajectoryWithRewPair], np.ndarray]:
...
+            custom_logger: Where to log to; if None (default), creates a new logger.
+            querent_kwargs: Keyword arguments passed to the querent.
+        """
+        super().__init__(wait_for_user, rng, custom_logger)
+        querent_kwargs = querent_kwargs if querent_kwargs else {}
+        self.querent = RESTQuerent(
+            collection_service_address=collection_service_address,
+            **querent_kwargs,
+        )
+        self.query_endpoint = collection_service_address + "/preferences/query/"
```
Comment (rk1a) [inline src/imitation/algorithms/preference_comparisons.py]: Include endpoint in address.

File: src/imitation/algorithms/preference_comparisons.py (hunk 10 above)
Comment (mschweizer) [reply src/imitation/algorithms/preference_comparisons.py]: Implemented the gatherer refactoring

File: src/imitation/algorithms/preference_comparisons.py
Comment (mschweizer) [reply src/imitation/algorithms/preference_comparisons.py:1]: Done

**Pull request merged**